### PR TITLE
Emma/optimize reanchor

### PR DESCRIPTION
**Pull request by emmaling27**
This PR improves the algorithm for `reanchor` - see https://github.com/paritytech/polkadot/issues/4489 for more details.

Instead of finding the common ancestor, appending junctions, and simplifying, we instead handle the cases where `target` and `self` are not on the same path directly by adding parents and `ancestry` as a junction if needed, and then in the case of `target` and `self` on the same branch, walk down both paths until they split, return the number of parents as the number of remaining junctions to reach `target` and the remaining `Junctions` to reach `self` as the new path. I increased test coverage, too.

### CODE REVIEW

File: xcm/src/v1/multilocation.rs [hunk 1]
```diff
@@ -329,18 +329,40 @@ impl MultiLocation {
...
+				// In the case where the paths are the same all the way to the target,
+				// remove all the junctions on the shared path
+				for _ in 0..target.interior.len() {
+					self.parents = 0;
+					self.take_first_interior();
+				}
+			}
+		} else if target.parents > self.parents {
+			// Handle when the target has more parents so we have to go through ancestry
+			self.parents = target.len() as u8 - 1;
```
Comment (KiChjang) [inline xcm/src/v1/multilocation.rs:358]: This is assuming that the `target` only has 1 more parent than `self`, which isn't always the case. I expect to see a loop here in fact.

File: xcm/src/v1/multilocation.rs [hunk 2]
```diff
@@ -329,18 +329,40 @@ impl MultiLocation {
...
+			self.interior
+				.push_front(ancestry.interior.first().unwrap_or(&Junction::OnlyChild).clone())
```
Comment (KiChjang) [inline xcm/src/v1/multilocation.rs:360]: This is only grabbing the first junction off of the ancestry? That doesn't seem correct to me. Could you try creating test cases where the `ancestry` isn't just `Parachain(2000)`? For example, imagine we're on a smart contract, and so the `ancestry` could look something like `Parachain(2022)/PalletInstance(50)/GeneralIndex(32)`.

File: xcm/src/v1/multilocation.rs [hunk 3]
```diff
@@ -329,18 +329,40 @@ impl MultiLocation {
...
+				.map_err(|_| ())?;
+		} else if target.parents < self.parents {
+			// Handle when the id has more parents so we must go all the way up the target path
+			self.parents = self.parents + target.interior().len() as u8;
```
Comment (KiChjang) [inline xcm/src/v1/multilocation.rs:364]: If `self`/`id` has more parents, then it doesn't sound like it's simply an addition here, I'd expect to see something like `(self.parents - target.parents) + target.interior().len() as u8`. The `self.parents - target.parents` part is the path used by the `target` to reach the part of `ancestry` in which `self`/`id` actually branches off from and starts concatenating interior junctions.

File: xcm/src/v1/multilocation.rs (hunk 1 above)
Comment (emmaling27) [reply xcm/src/v1/multilocation.rs:358]: I was actually wondering about this but then I forgot to leave a comment! Thanks!

**Comment by KiChjang** [comment]
@emmaling27 I appreciate the effort in trying to understand how `MultiLocation`s and the reanchoring algorithm works! This PR however still needs more work.

On a top-level view, I'd like to keep things a bit more declarative as possible, i.e. giving labels and names to algorithms and procedures, given the difficultly of understanding `MultiLocation` operations. What it still looks like to me, regardless of how `self.parents` compares with `target.parents`, is that the code is still trying to look for the common ancestor between `self` and `parent`, so there's no reason not to create a new internal function called `mutate_to_common_ancestor` or something similar just for this specific step in the algorithm. From there on, you can then append the remaining junctions to reach the final `MultiLocation`.

File: xcm/src/v1/multilocation.rs (hunk 2 above)
Comment (emmaling27) [reply xcm/src/v1/multilocation.rs:360]: Good point, I made a false assumption that the ancestry only had once junction, which made it hard to reason about the multiple parents as well. I'm confused at what the multilocation ancestry is relative to - will ancestry ever have parents > 0? Are `self` and `target` MultiLocations relative to the first junction in `ancestry` or the last one?

Comment (emmaling27) [reply xcm/src/v1/multilocation.rs:360]: Also, I'm not sure what we would want to default to if there isn't an anything in `ancestry.interior`. I feel like we should error here, but the way I'm familiar with doing error handling in rust would require me to change the fn signature to be `Result<(), Error>` and I haven't seen that elsewhere in this file. How does Result<(), ()> work? From my perspective, this gives you no information on how something failed, right?

Comment (KiChjang) [reply xcm/src/v1/multilocation.rs:360]: > will ancestry ever have parents

No, as I have mentioned in the issue, `ancestry` is not supposed to have parents, and in fact in an upcoming XCMv3 PR, we have specifically typed it as `InteriorMultiLocation`, a change from just `MultiLocation`.

The default of using `OnlyChild` looks fine to me.

**Pull request closed without merging**